In [37]:
import sklearn
from sklearn import datasets  
from pprint import pprint
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
test = sklearn.datasets.load_files("C:/Users/Lenovo Demo/Desktop/Data_Analytic/Text-Classification-master/txt_sentoken/Testing_data", description=None, categories=None, load_content=True, shuffle=True, encoding='utf-8', decode_error='ignore', random_state=0)
train = sklearn.datasets.load_files("C:/Users/Lenovo Demo/Desktop/Data_Analytic/Text-Classification-master/txt_sentoken/Training_data", description=None, categories=None, load_content=True, shuffle=True, encoding='utf-8', decode_error='ignore', random_state=0)

In [44]:
train.target_names

['neg', 'pos']

In [33]:
text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

text_clf.fit(train.data, train.target)

Pipeline(memory=None,
         steps=[('vect',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabulary=None)),
                ('tfidf',
                 TfidfTransformer(norm='l2', smooth_idf=True,
                                  sublinear_tf=False, use_idf=True)),
                ('clf',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [34]:
predicted_train = text_clf.predict(train.data)

print('We got an accuracy of',np.mean(predicted_train == train.target)*100, '% over the train data.')

We got an accuracy of 96.6875 % over the train data.


In [38]:
print("Classification report for testing data :-")
print(classification_report(train.target,predicted_train ))

Classification report for testing data :-
              precision    recall  f1-score   support

           0       0.95      0.99      0.97       800
           1       0.99      0.94      0.97       800

    accuracy                           0.97      1600
   macro avg       0.97      0.97      0.97      1600
weighted avg       0.97      0.97      0.97      1600



precision_0 = 0.95 ถ้าเลื่อกข้อความให้เป็น neg จะมีโอกาสถูก 95%
precision_1 = 0.99 ถ้าเลื่อกข้อความให้เป็น pos จะมีโอกาสถูก 99%

recall_0 = 0.99 จับข้อความที่เป็น neg ได้ 99% (มีโอกาสที่ neg จะหลุดไปเป็น pos อยู่ 1%)
recall_1 = 0.94 จับข้อความที่เป็น pos ได้ 94% (มีโอกาสที่ pos จะหลุดไปเป็น neg อยู่ 4%)

F1-score = 0.97 เฉลี่ยความถูกต้อง precision และ recall ที่อัตราส่วนเท่ากันได้ 97%

accuracy = 0.97 การแยกข้อความมีโอกาสถูก 97%

In [36]:
predicted_test = text_clf.predict(test.data)

print('We got an accuracy of',np.mean(predicted_test == test.target)*100, '% over the test data.')

We got an accuracy of 80.5 % over the test data.


In [39]:
print("Classification report for testing data :-")
print(classification_report(test.target,predicted_test ))

Classification report for testing data :-
              precision    recall  f1-score   support

           0       0.76      0.89      0.82       200
           1       0.87      0.72      0.79       200

    accuracy                           0.81       400
   macro avg       0.81      0.80      0.80       400
weighted avg       0.81      0.81      0.80       400



precision_0 = 0.76 ถ้าเลื่อกข้อความให้เป็น neg จะมีโอกาสถูก 76%
precision_1 = 0.87 ถ้าเลื่อกข้อความให้เป็น pos จะมีโอกาสถูก 87%

recall_0 = 0.89 จับข้อความที่เป็น neg ได้ 89% (มีโอกาสที่ neg จะหลุดไปเป็น pos อยู่ 11%)
recall_1 = 0.72 จับข้อความที่เป็น pos ได้ 72% (มีโอกาสที่ pos จะหลุดไปเป็น neg อยู่ 28%)

F1-score_0 = 0.97 เฉลี่ยความถูกต้อง precision_0 และ recall_0 ที่อัตราส่วนเท่ากันได้ 82%
F1-score_1 = 0.97 เฉลี่ยความถูกต้อง precision_1 และ recall_1 ที่อัตราส่วนเท่ากันได้ 82%

accuracy = 0.81 การแยกข้อความมีโอกาสถูก 81%